### PR TITLE
[core-v2-testnet] Deploy term

### PR DESCRIPTION
**Pull request by cashd**
### Scripts
Deploy script to deploy every contract needed to deploy a term. using mock contracts from the protocol-v2 repo

### Deploy functions
- forwarder factory
- mock token
- mock yearn vault
- mock yield adapter
- term vault
- vault

### Creation functions
- createTerm

### Utils
- time
- getCurrentBlockTimestamp
- getTokenIds (pt/yt)
- validateAddresses

### CODE REVIEW

File: apps/core-v2-testnet/scripts/deploy.ts [hunk 1]
```diff
@@ -26,6 +32,44 @@ async function main() {
     maxReserve,
     signer.address,
   );
+
+  // deploy mock yearn vaults
+  // todo see how many decimals is usdc token
+  const usdcToken = await deployMockToken(signer, "USDC", "USDC", 18);
```
Comment (sentilesdal) [inline apps/core-v2-testnet/scripts/deploy.ts]: nitpick USDC is 6 decimals ;)

File: apps/core-v2-testnet/scripts/deploy.ts [hunk 2]
```diff
@@ -26,6 +32,44 @@ async function main() {
...
+  );
+  console.log("yield adapter");
+
+  // give token to signer
+  usdcToken.connect(signer).mint(signer.address, 7e6);
+  // set allowance for yield adapter
+  usdcToken.approve(mockYieldAdapter.address, 12e6);
+  // create new term with underlying tokens
+  const currentTimestamp = await getCurrentBlockTimestamp(ethers.provider);
+  const start = BigNumber.from(currentTimestamp + ONE_MINUTE);
```
Comment (sentilesdal) [inline apps/core-v2-testnet/scripts/deploy.ts:51]: any start in the future will default to the present.  you can backdate though.

**Review by sentilesdal — COMMENTED** [review]
Prolly better to not call the Term a vault.  It technically is because it holds _some_ funds but mostly its job is to move those funds to another vault or pool.

File: apps/core-v2-testnet/src/deploy/index.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+// barrell
```
[the remark this answers is not in the document]
Comment (cashd) [reply apps/core-v2-testnet/src/deploy/index.ts]: haha, I removed this on later commit

File: apps/core-v2-testnet/src/deploy/deployTerm.ts [hunk 4]
```diff
@@ -1 +1,34 @@
-export async function deployTerm(): Promise<void> {}
+import {
+  ERC4626Term,
+  ERC4626Term__factory,
+} from "@elementfi/core-v2-typechain";
+import { Signer, BigNumberish } from "ethers";
+import { ethers } from "hardhat";
+import { throwInvalidAddressError } from "src/errors";
+
+export async function deployTerm(
```
[the remark this answers is not in the document]
Comment (cashd) [reply apps/core-v2-testnet/src/deploy/deployTerm.ts]: I renamed it at a later commit to, `deployTermVault`, thoughts?

File: yarn.lock [hunk 5]
```diff
@@ -28493,7 +28511,7 @@ snapdragon@^0.8.1:
     source-map-resolve "^0.5.0"
     use "^3.1.0"
 
-snarkjs@0.4.10, snarkjs@^0.4.10:
+snarkjs@^0.4.10:
```
[the remark this answers is not in the document]
Comment (cashd) [reply yarn.lock:28514]: ooops, my yarn automatically does this

File: apps/core-v2-testnet/src/utils/time.ts [hunk 6]
```diff
@@ -0,0 +1,4 @@
+export const ONE_SECOND = 1;
```
[the remark this answers is not in the document]
Comment (cashd) [reply apps/core-v2-testnet/src/utils/time.ts:1]: yes there are! removing this version

File: apps/core-v2-testnet/scripts/deploy.ts [hunk 7]
```diff
@@ -1,21 +1,66 @@
...
 
-  await lock.deployed();
+  // deploy mock yearn vaults
+  // todo see how many decimals is usdc token
 
-  Logger.successfulDeploy("Lock", lock);
+  // Deploy mock vault
+  const mockYearnVault = await deployMockYearnVault(signer, usdcToken.address);
+
+  const governanceAddress = ethers.constants.AddressZero;
```
Comment (DannyDelott) [inline apps/core-v2-testnet/scripts/deploy.ts:33]: nit: Add a small comment why we need this, ie: why do yield adapters need the governanceAddress?

File: apps/core-v2-testnet/scripts/deploy.ts [hunk 8]
```diff
@@ -1,21 +1,66 @@
...
+    signer,
+    mockYearnVault.address,
+    governanceAddress,
+    linkHash,
+    forwarderFactory.address,
+    usdcToken.address,
+  );
+
+  // give tokens to signer
+  usdcToken.connect(signer).mint(signer.address, 7e6);
```
Comment (DannyDelott) [inline apps/core-v2-testnet/scripts/deploy.ts:45]: nit: move amounts to a named variable

File: apps/core-v2-testnet/scripts/deploy.ts [hunk 9]
```diff
@@ -1,21 +1,66 @@
...
+  // set allowance for yield adapter
+  usdcToken.approve(mockYieldAdapter.address, 12e6);
```
Comment (DannyDelott) [inline apps/core-v2-testnet/scripts/deploy.ts:47]: Who's the spender? Update comment might be: "set allowance for yield adapter to spend signer's usdc"

File: apps/core-v2-testnet/src/helpers/createTerm.ts [hunk 10]
```diff
@@ -0,0 +1,31 @@
+import { Term } from "@elementfi/core-v2-typechain";
+import { BigNumber, Signer } from "ethers";
+import { validateAddresses } from "src/utils/validateAddresses";
+
+// Creates new term within existing term contract
+export async function createTerm(
+  signer: Signer,
+  term: Term,
+  start: BigNumber, // unix timestamp seconds
+  expiry: BigNumber, // unix timestamp seconds
```
Comment (DannyDelott) [inline apps/core-v2-testnet/src/helpers/createTerm.ts:10]: nit: unix timestamp seconds can just be a `number` argument. We should convert it to BigNumber when we call `lock()` instead.

File: apps/core-v2-testnet/src/utils/getTokenIds.ts [hunk 11]
```diff
@@ -0,0 +1,17 @@
+// Utils to get PT and YT token ids
+// Token ids reference the token values in the MultiToken contract
+
+import { BigNumber, BigNumberish } from "ethers";
+
+export function getPTId(expiry: BigNumberish): BigNumber {
+  return BigNumber.from(expiry);
+}
```
Comment (DannyDelott) [inline apps/core-v2-testnet/src/utils/getTokenIds.ts:8]: Hmm, passing ids around as BigNumbers feels weird to me, can we use strings or numbers instead?

File: apps/core-v2-testnet/src/utils/validateAddresses.ts [hunk 12]
```diff
@@ -0,0 +1,11 @@
+import { ethers } from "hardhat";
+import { throwInvalidAddressError } from "src/errors";
+
+export function validateAddresses(addresses: string[]): void | never {
```
Comment (DannyDelott) [inline apps/core-v2-testnet/src/utils/validateAddresses.ts:4]: this is neat, add a TODO to move it into @elementfi/base?

**Review by DannyDelott — APPROVED** [review]
Nice! Couple of small things

File: apps/core-v2-testnet/scripts/deploy.ts (hunk 7 above)
Comment (cashd) [reply apps/core-v2-testnet/scripts/deploy.ts:33]: it get's passed to a term contract through inheritance. Yield adapters are terms basically but extend an additional interface

Comment (cashd) [reply apps/core-v2-testnet/scripts/deploy.ts:33]: Terms have an owner constructor param

File: apps/core-v2-testnet/src/utils/validateAddresses.ts (hunk 12 above)
Comment (cashd) [reply apps/core-v2-testnet/src/utils/validateAddresses.ts:4]: nice suggestion

**Pull request merged**